core/src/main/java/net/opentsdb/data/types/numeric/aggregators/ArrayAggregatorUtils.java

```diff
@@ -1,2 +1,211 @@
+    OUT_OF_BOUNDS,
+
+    /** At least one value was written within the boundary. */
+    SUCCESS
+  }
+
+  /**
+   * Writes the given time series into the proper location of the aggregation
+   * array given that the time series has a {@link NumericArrayType} iterator.
+   * The location in the aggregator array is determined by the timestamp of the
+   * numeric array value so make sure that is correct when using this method.
+   * <b>Invariants:</b>
+   * <ul>
+   *   <li>The aggEnd must be greater than the aggStart.</li>
+   *   <li>AggEnd must be at least one interval greater than aggStart.</li>
+   *   <li>The time series array iterator must have the same interval as the
+   *   aggregator. Check that before using this method otherwise the aggregator
+   *   will have garbage.</li>
+   * </ul>
+   * @param agg The non-null aggregator to write into.
+   * @param aggStart The non-null starting timestamp of the aggregator array.
+   * @param aggEnd The non-null ending timestamp of the aggregator array,
+   *               exclusive.
+   * @param interval The non-null interval between values in the two arrays.
+   * @param timeseries The non-null time series to pull an iterator from.
+   * @return A non-null state based on the aggregation.
+   */
+  public static AccumulateState accumulateInAggregatorArray(
+          final NumericArrayAggregator agg,
+          final TimeStamp aggStart,
+          final TimeStamp aggEnd,
+          final TemporalAmount interval,
+          final TimeSeries timeseries) {
+
+    final Optional<TypedTimeSeriesIterator<? extends TimeSeriesDataType>> op =
+            timeseries.iterator(NumericArrayAggregator.TYPE);
+    if (!op.isPresent()) {
+      return AccumulateState.NOT_PRESENT;
+    }
+    final TypedTimeSeriesIterator<? extends TimeSeriesDataType> iterator = op.get();
+    if (!iterator.hasNext()) {
+      return AccumulateState.NO_VALUE;
+    }
+
+    final TimeSeriesValue<NumericArrayType> value =
+            (TimeSeriesValue<NumericArrayType>) iterator.next();
+    if (value.timestamp().compare(TimeStamp.Op.GTE, aggEnd)) {
+      return AccumulateState.OUT_OF_BOUNDS;
+    }
+
+    // TODO - only handling seconds for now. need ms and nanos someday.
+    TimeStamp currentTs = value.timestamp().getCopy();
+    int aggIndex = (int) ((value.timestamp().epoch() - aggStart.epoch()) /
+            interval.get(ChronoUnit.SECONDS));
+    int arrayIndex = value.value().offset();
+    // make sure we move to the start of the agg index if we have data before
+    // the interval.
+    while (aggIndex < 0 && arrayIndex < value.value().end() &&
+            currentTs.compare(TimeStamp.Op.LT, aggEnd)) {
+      currentTs.add(interval);
+      ++aggIndex;
+      ++arrayIndex;
+    }
+    if (aggIndex < 0) {
+      return AccumulateState.OUT_OF_BOUNDS;
+    }
+
+    int wrote = 0;
+    if (value.value().isInteger()) {
+      long[] array = value.value().longArray();
+      while (arrayIndex < value.value().end() &&
```

Comment:
This common part can be moved out of if else.

Reply:
I should test it but I'm trying to help tighten the loop for cache locality and avoid another branch. It could be compiled out at run time though, I'd have to see.